types/module/server/manager.go

```diff
@@ -0,0 +1,149 @@
+package server
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkmodule "github.com/cosmos/cosmos-sdk/types/module"
+	gogogrpc "github.com/gogo/protobuf/grpc"
+
+	"github.com/regen-network/regen-ledger/types/module"
+)
+
+type Manager struct {
+	sdkmodule.Manager
+	baseApp          *baseapp.BaseApp
+	cdc              *codec.ProtoCodec
+	keys             map[string]ModuleKey
+	router           *router
+	requiredServices map[reflect.Type]bool
+}
+
+func NewManager(baseApp *baseapp.BaseApp, cdc *codec.ProtoCodec) *Manager {
+	return &Manager{
+		baseApp: baseApp,
+		cdc:     cdc,
+		keys:    map[string]ModuleKey{},
+		router: &router{
+			handlers:         map[string]handler{},
+			providedServices: map[reflect.Type]bool{},
+			antiReentryMap:   map[string]bool{},
+		},
+	}
+}
+
+func (mm *Manager) RegisterModules(modules []module.Module) error {
+	for _, mod := range modules {
+		// check if we actually have a server module, otherwise skip
+		serverMod, ok := mod.(Module)
+		if !ok {
+			continue
+		}
+
+		serverMod.RegisterInterfaces(mm.cdc.InterfaceRegistry())
+	}
+
+	for _, mod := range modules {
+		// check if we actually have a server module, otherwise skip
+		serverMod, ok := mod.(Module)
+		if !ok {
+			continue
+		}
```

Comment:
Maybe i'm missing the rationale.. but it looks like this second for loop is uncessary, and should be combined with the for loop for lines 39-47? Lmk if i'm missing something and if there is a real need for running the `serverMod.RegisterInterfaces()` calls in a separate forloop before.

Reply:
Update: The separate forloop is prefered so that we can first propagate errors on interface registry failures before trying to do any other logic. A comment to explain this would be nice.